feat: Clearer 'Packaging components' preview and modified the maximum size of the autocomplete tab.

thomas-algo opened the pull request:
### What
<!-- description of the PR -->
- Updated the preview of the package in the 'Packaging components ' tab (cf two first screenshots)
    - Clearer
    - More coherent with what is shown on the product page.
    - Updated the getTitle() and created getSubTitle() function from packages/smooth_app/lib/pages/product/edit_new_packagings.dart 

- Autocomplete now gives hints before anything is typed. (removed from this PR, implemented in #3750)
    - The autocomplete now gives the most common answers for the field before user start typing. (Useful for packaging material, recycling...) (cf third screenshot)

- Modified the maximum size of the autocomplete tab.
    -  The autocomplete tab was unreadable if the text field was too close to the keyboard, even if the page was scrolled afterward. (cf two last screenshots) 
    - Fixing _AssertionError 'maxOptionsHeight >= 0': is not true: The old formula could give negative values.

### Screenshot
<!-- Insert a screenshot to provide visual record of your changes, if visible -->
- Old preview of the package in the 'Packaging components ' tab
![image](https://user-images.githubusercontent.com/105145857/221690480-6d106387-6a65-4c76-9b8a-dfe33ab0d2b5.png)

- New preview of the package in the 'Packaging components ' tab
![image](https://user-images.githubusercontent.com/105145857/221690636-5855cf95-49e3-4923-a6c7-9b4cd6c75d27.png)

- Autocomplete now gives hints before anything is typed
![image](https://user-images.githubusercontent.com/105145857/221696034-cdae1a59-8362-41d5-bf1e-32c09fbdb259.png)


- Old behavior of the autocomplete max size

![image](https://user-images.githubusercontent.com/105145857/221695543-5775b869-5742-42ad-9439-cbe1a5df0df9.jpeg)

Stays small even when moved:
![image](https://user-images.githubusercontent.com/105145857/221695243-2a007b88-2f52-4bd2-a7a6-95285aaf06aa.jpeg)

thomas-algo committed:
Updated the preview of the package in the 'Packaging components ' tab.

Preview is now clearer and more coherent with what is shown on the product page.

thomas-algo committed:
Autocomplete now gives hints before anything is typed & maxOptionsHeight fixed

- The autocomplete now gives the most common answers for the field before user start typing. (Useful for packaging material, recycling...)

- Fixing _AssertionError 'maxOptionsHeight >= 0': is not true:
      The old formula could give negative values. It could also create an unreadable autocomplete if the field was too close to the keyboard.

thomas-algo committed:
Merge branch 'develop' into edit_packaging

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_new_packagings_helper.dart:
```diff
@@ -75,23 +75,36 @@ class EditNewPackagingsHelper {
-    addIfNotEmpty(controllerWeight.text);
     addIfNotEmpty(controllerQuantity.text);
 
+    if (controllerMaterial.text.isNotEmpty & controllerWeight.text.isNotEmpty) {
+      result.add('(${controllerMaterial.text} : ${controllerWeight.text}g)');
+    }
+
+    else if (controllerMaterial.text.isNotEmpty ) { // Therefore controllerWeight.text is empty
+          result.add('(${controllerMaterial.text})');
+    }
+
```
```suggestion
    if (controllerMaterial.text.isNotEmpty) {
      if (controllerWeight.text.isNotEmpty) {
        result.add('(${controllerMaterial.text} : ${controllerWeight.text}g)');
      } else {
        result.add('(${controllerMaterial.text})');
      }
    }
```

So you don't display the weight if the material is not specified, which is a bit strange.

thomas-algo replied:
Yes, this is the same behavior as in the final product page. However, it might be interesting to have the mass in both cases (Nevertheless, the case where 'material' is empty and 'weight' is filled is quit rare).

monsieurtanuki replied:
Again, I don't know where the changed display was suggested. And how rare the "the case where 'material' is empty and 'weight' is filled".
I would suggest at least to consider this very rare case with an additional test: "if weight is not empty and material is empty, let's display only weight".
By the way in English there's no space before a `:`. In French it's an unbreakable space.

thomas-algo replied:
```suggestion
if (controllerMaterial.text.isNotEmpty & controllerWeight.text.isNotEmpty) {
      result.add('(${controllerMaterial.text}: ${controllerWeight.text}g)');
    }

    if (controllerMaterial.text.isNotEmpty & controllerWeight.text.isEmpty) {
          result.add('(${controllerMaterial.text})');
    }
  
    if (controllerMaterial.text.isEmpty & controllerWeight.text.isNotEmpty) {
          result.add('(${controllerWeight.text}g)');
    }
    // And nothing is added if they are both empty.
```

Thanks for your advises, here is a version taking this case into account. I am not sure the display change was suggested elsewhere.

stephanegigandet replied:
> By the way in English there's no space before a `:`. In French it's an unbreakable space.

@monsieurtanuki: on the server we have a translatable string 'sep' which is an empty string in English, and a space in French and other languages. Is this something already handled in some way in the app? I didn't see something equivalent in the lib/l10n/app_en.arb file

monsieurtanuki replied:
@stephanegigandet We don't have this `sep` in Smoothie.
Could you list all languages that do have a space, e.g. taken from the server?

stephanegigandet replied:
@monsieurtanuki I just checked, we only do it for French.

```
=head2 separator_before_colon( $l )

=head3 Arguments

=head4 language code $l

In some languages like French, colons have a space before them.
e.g. "Valeur : 500" in French, "Value: 500" in English

This function returns a non-breaking space character for those languages.

=cut

sub separator_before_colon ($l) {

	if ($l eq 'fr') {
		return "\N{U+00A0}";
	}
	else {
		return '';
	}
}
```

Maybe we can put a similar function in Smoothie, it's likely to be useful in other places.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_new_packagings_helper.dart:
```diff
@@ -75,23 +75,36 @@ class EditNewPackagingsHelper {
       return null;
     }
-    return result.join(' ');
+    return result.join('');
```
I cannot see the point of adding spaces everywhere so that you can remove the space from the `join`, but why not.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/simple_input_text_field.dart:
```diff
@@ -40,19 +40,16 @@ class SimpleInputTextField extends StatelessWidget {
                     language: ProductQuery.getLanguage()!,
-                    limit: 1000000, // lower max count on the server anyway
-                    input: value.text.trim(),
+                    limit: 15,
```
Why `15`?

thomas-algo replied:
The limit is 400 on the server. The autocomplete tab doesn't need to be that long. Moreover, if the maximum API limit on the server was increased, nothing would limit the number of items in the suggestions.

monsieurtanuki replied:
Therefore your answer is: "`15` is the best compromise according to me". Please add a comment about it.

thomas-algo replied:
Yes, I believe that it is the best compromise, it is unnecessary to have more that 15 suggestions as approximately 5 can be seen at once on screen without scrolling.

M123-dev replied:
15 sounds reasonable, but yes a comment would be good 👍🏼

thomas-algo replied:
Done!

monsieurtanuki commented on packages/smooth_app/lib/pages/product/simple_input_text_field.dart:
```diff
@@ -40,19 +40,16 @@ class SimpleInputTextField extends StatelessWidget {
                     return <String>[];
                   }
-
+                  
```
```suggestion

```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/simple_input_text_field.dart:
```diff
@@ -100,17 +97,7 @@ class SimpleInputTextField extends StatelessWidget {
-                        (DefaultTextStyle.of(context).style.fontSize ?? 0) -
-                        // Elevation
-                        4.0,
+                    maxOptionsHeight: screenHeight/3,
```
Just hoping it works.

thomas-algo replied:
You can see the result on the third screenshot

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_new_packagings_helper.dart:
```diff
@@ -75,23 +75,36 @@ class EditNewPackagingsHelper {
 
+  /// Returns the packaging subtitle from the controllers
+String getSubTitle() {
+    return controllerRecycling.text;
```
Says who?

thomas-algo replied:
I moved the packaging recycling info to the subtitle of the widget for readability (cf first screenshot). This function generate this subtitle.

monsieurtanuki reviewed (commented):
¡Hi @thomas-algo!
Please have a look at my comments.
I approve some of your UI/UX changes, but I cannot see the related issue.

stephanegigandet commented on packages/smooth_app/lib/pages/product/simple_input_text_field.dart:
```diff
@@ -40,19 +40,16 @@ class SimpleInputTextField extends StatelessWidget {
                 optionsBuilder: (final TextEditingValue value) async {
                   final String input = value.text.trim();
 
-                  if (input.isEmpty) {
```
This will be great for packaging shapes and packaging materials, in order to have the most popular suggestions available without having to type the first letter. But we probably should not do it for other tagType values than packaging_shapes and packaging_materials, as it probably would be confusing for things like categories and labels. Especially as currently we return suggestions in alphabetic order for them.

stephanegigandet replied:
Also addressed by @monsieurtanuki in #3750

thomas-algo replied:
The solution provided in #3750 is indeed better, I removed mine from this PR.

stephanegigandet commented:
Hi @thomas-algo, thanks a lot for this PR. This will definitely ease a lot the work to enter packaging data.

As @monsieurtanuki suggests, it's best to first file issues about what you intend to change, so that others can comment, make suggestions etc. I'll file the corresponding issues for this PR.

thomas-algo committed:
Removed Autocomplete now gives hints before anything is typed.

@monsieurtanuki as addressed the issue in #3750 without disturbing old default behavior of the widget.

monsieurtanuki committed:
Merge branch 'develop' into edit_packaging

M123-dev commented:
Phew, that were a lot of comments in a really short time.

The automated tests are currently failing due to file formatting, please run `flutter format .` in packages/smooth_app
This will fix them :D

thomas-algo committed:
Comment explaining the limit of suggestions.

thomas-algo committed:
Taking into account the case 'Material empty' and 'Weight not empty' & format

thomas-algo committed:
Removing unused variables

This variables where previously used in the formula of maxOptionsHeight

monsieurtanuki committed:
Merge branch 'develop' into edit_packaging

stephanegigandet reviewed (approved):
Looks good to me. Can we merge? cc @monsieurtanuki @M123-dev

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_new_packagings_helper.dart:
```diff
@@ -75,21 +75,39 @@ class EditNewPackagingsHelper {
-    addIfNotEmpty(controllerWeight.text);
     addIfNotEmpty(controllerQuantity.text);
 
+    if (controllerMaterial.text.isNotEmpty & controllerWeight.text.isNotEmpty) {
+      result.add('(${controllerMaterial.text}: ${controllerWeight.text}g)');
+    }
+
+    if (controllerMaterial.text.isNotEmpty & controllerWeight.text.isEmpty) {
+      result.add('(${controllerMaterial.text})');
+    }
+
+    if (controllerMaterial.text.isEmpty & controllerWeight.text.isNotEmpty) {
+      result.add('(${controllerWeight.text}g)');
+    }
+    // And nothing is added if they are both empty.
+
```
```suggestion
    if (controllerMaterial.text.isNotEmpty) {
      if (controllerWeight.text.isNotEmpty) {
        result.add('(${controllerMaterial.text}: ${controllerWeight.text}g)');
      } else {
        result.add('(${controllerMaterial.text})');
      }
    } else if (controllerWeight.text.isNotEmpty) {
      result.add('(${controllerWeight.text}g)');
    }

```

monsieurtanuki committed:
Update packages/smooth_app/lib/pages/product/edit_new_packagings_helper.dart

monsieurtanuki commented:
I cannot say that I got all my answers but I see nothing too problematic in this PR.
For the record, in `dart` (as in many languages) the boolean "AND" operator is `&&`, not `&`.
